File: frontend/src/types/task.ts
export type TaskStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';

export type StemType = 'vocals' | 'instrumental' | 'drums' | 'bass' | 'other';

export type ModelType = 'demucs' | 'demucs_ft' | 'bs_roformer' | 'melband_roformer';

export interface SeparateResponse {
  task_id: string;
  status: TaskStatus;
  credit_consumed: number;
  credits_remaining: number;
}

export interface TaskDetail {
  id: string;
  status: TaskStatus;
  model_used: string;
  credit_consumed: number;
  duration_seconds: number | null;
  original_filename: string;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
  download_urls: Record<string, string> | null;
  error_log: string | null;
}

export interface TaskHistoryResponse {
  items: TaskDetail[];
  total: number;
  page: number;
  pages: number;
  has_next: boolean;
}

export interface TaskHistoryParams {
  page?: number;
  limit?: number;
  status?: TaskStatus;
  sort_by?: string;
  order?: 'asc' | 'desc';
}

File: backend/app/schemas/task.py
import uuid
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, ConfigDict


# ── Separation request ────────────────────────────────────────────────────────

class SeparateRequest(BaseModel):
    model: str = "demucs"


# ── Task responses ────────────────────────────────────────────────────────────

class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: str
    model_used: Optional[str] = None
    credit_consumed: int
    duration_seconds: Optional[float] = None
    original_filename: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None


class DownloadURLs(BaseModel):
    vocals: Optional[str] = None
    instrumental: Optional[str] = None


class TaskDetailResponse(TaskResponse):
    download_urls: Optional[DownloadURLs] = None
    error_log: Optional[str] = None


class TaskHistoryResponse(BaseModel):
    items: List[TaskResponse]
    total: int
    page: int
    pages: int
    has_next: bool


# ── Separation submit response ────────────────────────────────────────────────

class SeparateResponse(BaseModel):
    task_id: uuid.UUID
    status: str
    credit_consumed: int
    credits_remaining: int


# ── Credit schemas ────────────────────────────────────────────────────────────

class CreditBalanceResponse(BaseModel):
    balance: int
    total_consumed: int
    total_purchased: int


class CreditTransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    amount: int
    type: str
    task_id: Optional[uuid.UUID] = None
    description: Optional[str] = None
    created_at: datetime


class CreditTransactionHistoryResponse(BaseModel):
    items: List[CreditTransactionResponse]
    total: int
    page: int
    pages: int
    has_next: bool


class PurchaseRequest(BaseModel):
    amount: int  # number of credits to purchase (mock payment)


class RedeemRequest(BaseModel):
    code: str
